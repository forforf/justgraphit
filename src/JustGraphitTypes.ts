
// Not perfect, but ok for now. Maybe improve if typescript supports regexes
export type ISODatetimeType = `${number}-${number}-${number}T${number}:${number}:${number}.${number}Z`;

export type GraphName = string;

export type DateOrIso = Date | ISODatetimeType;

// individual data points
export type JustGraphitEntry = {
  number: number;
  datetime: ISODatetimeType;
  dt?: Date;
}

export type JustGraphitStore = Record<GraphName, JustGraphitEntry[]>

export type PassthroughFunction = (...args: unknown[]) => unknown;

export type RefSvg = SVGSVGElement | null;

export type D3Select = (ref: RefSvg) => PassthroughFunction;

// Functions that handle graphs by graphName (changing graphs, deleting graphs, etc)
export type GraphHandler = (graphName: GraphName) => void;

export type GraphEntryHandler = (graphName: GraphName, numberInput: string) => void;