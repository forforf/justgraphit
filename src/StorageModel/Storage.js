import IsEmpty from 'lodash.isempty';
import History from './History';


// ToDo Needs testing
class LocalStorageWrapper {
  constructor(mainKey='__justgraphit__') {
    this.mainKey = mainKey;
  }

  load = () => {
    return JSON.parse(localStorage.getItem(this.mainKey));
  };

  save = (value) =>  {
    localStorage.setItem(this.mainKey, JSON.stringify(value));
  };
}

class Storage {

  constructor(initialObj) {
    const initialObject = IsEmpty(initialObj) ? {} : initialObj;
    this.persistentStore = new LocalStorageWrapper();
    this.store = this.persistentStore.load();
    if (this.isEmpty()) {
      this.persistentStore.save(initialObject);
      this.store = this.persistentStore.load();
    }
    this.history = new History(2);
  }

  deleteObject = key => {
    const deleteRetVal = (delete this.store[key]);
    if (deleteRetVal) {
      this.updatePersistence();
    }
    return deleteRetVal;
  }
  
  getAll= () => this.store;

  getAllKeys = () => Object.keys(this.getAll());

  getInitialKey = () => Object.keys(this.store)[0];
  
  getLastKey = () => ( this.history.getLast() || this.getInitialKey() );

  // returns last object from history
  getLastStorageObject = () => {
    let obj = {};
    const key = this.getLastKey();
    obj[key] = this.load(key);
    return obj;
  };
  
  isEmpty = () => IsEmpty(this.store);

  load = key =>  this.store[key];

  save = (key, val) => {
    this.updateHistory(key);
    this.store[key] = val;
    this.updatePersistence();
    return key;
  };

  updateHistory = key => {
    if (key && key.length) {
      this.history.push(key);
    }
  }

  updatePersistence = () => {
    this.persistentStore.save(this.store);
  }
}

export default Storage;
