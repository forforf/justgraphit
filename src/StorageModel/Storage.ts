import IsEmpty from 'lodash.isempty';
import {
  GraphName,
  JustGraphitEntry,
  JustGraphitStore,
} from '../JustGraphitTypes';

// ToDo Needs testing:
// It's just an thin abstraction layer above a persistent store (local storage in this case)
// and should not have any JustGraphit dependencies
class LocalStorageWrapper {
  mainKey: string;

  constructor(mainKey = '__justgraphit__') {
    this.mainKey = mainKey;
  }

  load = () => {
    return JSON.parse(localStorage.getItem(this.mainKey) ?? '{}');
  };

  // We are saving an entire JSON object using mainKey
  // The saved JSON object should have keys of it's own to be
  // meaningful in this context.
  save = (value: unknown) => {
    localStorage.setItem(this.mainKey, JSON.stringify(value));
  };
}

class Storage {
  persistentStore: LocalStorageWrapper;
  store: JustGraphitStore;

  constructor(initialObj: JustGraphitStore) {
    const initialObject: JustGraphitStore = IsEmpty(initialObj)
      ? {}
      : initialObj;
    this.persistentStore = new LocalStorageWrapper();
    this.store = this.persistentStore.load();
    if (this.isEmpty()) {
      this.persistentStore.save(initialObject);
      this.store = this.persistentStore.load();
    }
  }

  deleteObject = (key: string): boolean => {
    const deleteRetVal = delete this.store[key] ?? '';
    if (deleteRetVal) {
      this.updatePersistence();
    }
    return deleteRetVal;
  };

  getAll = (): JustGraphitStore => this.store;

  getAllKeys = (): GraphName[] => Object.keys(this.getAll());

  getInitialKey = (): GraphName => Object.keys(this.store)[0];

  getInitialObject = (): Record<GraphName, JustGraphitEntry[]> => {
    return { [this.getInitialKey()]: this.store[this.getInitialKey()] };
  };

  isEmpty = (): boolean => IsEmpty(this.store);

  load = (key: string): JustGraphitEntry[] => this.store[key];

  save = (key: GraphName, val: JustGraphitEntry[]): GraphName => {
    this.store[key] = val;
    this.updatePersistence();
    return key;
  };

  updatePersistence = (): void => {
    this.persistentStore.save(this.store);
  };
}

export default Storage;
