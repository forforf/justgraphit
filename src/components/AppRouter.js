import React, {useState} from 'react';
import {HashRouter, Route} from 'react-router-dom';
import GraphObject from '../StorageModel/GraphObject';
import {AppBar} from '@mui/material';
import NavBar from './NavBar';
import Graphit from './Graphit';
import About from './About';
import Editor from './Editor';

function makePropless(Component, componentProps) {
  return (() => <Component {...componentProps} />)
}

function AppRouter({storage}) {
  const lastObject = storage.getLastStorageObject();

  // state
  const [graphObject, setGraphObject] = useState(GraphObject.fromPlainObject(lastObject));
  const [graphNameList, setGraphNameList] = useState(storage.getAllKeys());

  // CRUD-like handlers
  const changeSelectedGraph = graphName => {
    const newGraphData = storage.load(graphName);
    storage.updateHistory(graphName)
    const newGraphObject = new GraphObject(graphName, newGraphData)
    setGraphObject(newGraphObject);
  };

  const addNewNumber = (graphName, newNumberInput) => {
    const number = parseFloat(newNumberInput);
    const datetime = new Date().toISOString();
    const newEntry = { number, datetime };
    let newGraphData = [...graphObject.data, newEntry];
    // If we get a new graphName, create a new object
    if (graphName !== graphObject.name) {
      newGraphData = [newEntry];
      setGraphNameList([...graphNameList, graphName]);
    }
    const newGraphObject = new GraphObject(graphName, newGraphData);
    storage.save(graphName, newGraphData);
    setGraphObject(newGraphObject);
  }

  const deleteGraph = graphName => {
    if(window.confirm('Delete Graph?')) {
      storage.deleteObject(graphName);
      // Load last saved graph
      const lastName = storage.getLastKey()
      const lastData = storage.load(lastName);
      const newGraphObject = new GraphObject(lastName, lastData);
      setGraphObject(newGraphObject);
      setGraphNameList(storage.getAllKeys())
    }
  };

  // props
  const graphItProps = {
    graphObject,
    addNewNumber,
    changeSelectedGraph,
    deleteGraph,
    graphNameList
  }

  const editorProps = { graphObject, setGraphObject, saveToStorage: storage.save };

  return (
    <HashRouter>
      <AppBar title="JustGraphIt" />
      <NavBar />
      <Route exact path="/" component={makePropless(Graphit, {...graphItProps})} />
      <Route exact path="/about" component={About} />
      <Route exact path="/edit" component={makePropless(Editor, {...editorProps})} />
    </HashRouter>
  );
}

export default AppRouter
