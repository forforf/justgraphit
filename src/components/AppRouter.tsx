import React, { useState } from 'react';
import { HashRouter, Route } from 'react-router-dom';
import GraphObject from '../StorageModel/GraphObject';
import { AppBar } from '@mui/material';
import NavBar from './NavBar';
import Graphit, { GraphitProps } from './Graphit';
import About from './About';
import Editor, { EditorProps } from './Editor';
import {
  GraphEntryHandler,
  GraphHandler,
  ISODatetimeType,
  JustGraphitEntry,
} from '../JustGraphitTypes';
import Storage from '../StorageModel/Storage';

export type AppRouterProps = {
  storage: Storage;
};

function makePropless<T>(Component: React.ComponentType, componentProps: T) {
  function Propless(routerProps: unknown) {
    return <Component {...componentProps} {...routerProps} />;
  }

  Propless.displayName = `Propless(${Component.displayName || Component.name})`;
  return Propless;
}

function AppRouter({ storage }: AppRouterProps): JSX.Element {
  // const lastObject = storage.getLastStorageObject();
  const initialObject = storage.getInitialObject();

  // state
  const [graphObject, setGraphObject] = useState(
    GraphObject.fromPlainObject(initialObject)
  );
  const [graphNameList, setGraphNameList] = useState(storage.getAllKeys());

  // CRUD-like handlers
  const changeSelectedGraph: GraphHandler = (graphName) => {
    const newGraphData = storage.load(graphName);
    // storage.updateHistory(graphName)
    const newGraphObject = new GraphObject(graphName, newGraphData);
    setGraphObject(newGraphObject);
  };

  const addNewNumber: GraphEntryHandler = (graphName, newNumberInput) => {
    const number = parseFloat(newNumberInput);
    const datetime: ISODatetimeType =
      new Date().toISOString() as ISODatetimeType;
    const newEntry: JustGraphitEntry = { number, datetime };
    let newGraphData: JustGraphitEntry[] = [...graphObject.data, newEntry];
    // If we get a new graphName, create a new object
    if (graphName !== graphObject.name) {
      newGraphData = [newEntry];
      setGraphNameList([...graphNameList, graphName]);
    }
    const newGraphObject: GraphObject = new GraphObject(
      graphName,
      newGraphData
    );
    storage.save(graphName, newGraphData);
    setGraphObject(newGraphObject);
  };

  const deleteGraph: GraphHandler = (graphName) => {
    if (window.confirm('Delete Graph?')) {
      storage.deleteObject(graphName);
      // Load last saved graph
      const lastName = storage.getInitialKey();
      const lastData = storage.load(lastName);
      const newGraphObject = new GraphObject(lastName, lastData);
      setGraphObject(newGraphObject);
      setGraphNameList(storage.getAllKeys());
    }
  };

  // props
  const graphItProps = {
    graphObject,
    addNewNumber,
    changeSelectedGraph,
    deleteGraph,
    graphNameList,
  };

  const editorProps = {
    graphObject,
    setGraphObject,
    saveToStorage: storage.save,
  };

  return (
    <HashRouter>
      <AppBar title="JustGraphIt" />
      <NavBar />
      <Route
        exact
        path="/"
        component={makePropless<GraphitProps>(Graphit as React.ComponentType, {
          ...graphItProps,
        })}
      />
      <Route exact path="/about" component={About} />
      <Route
        exact
        path="/edit"
        component={makePropless<EditorProps>(Editor as React.ComponentType, {
          ...editorProps,
        })}
      />
    </HashRouter>
  );
}

export default AppRouter;
