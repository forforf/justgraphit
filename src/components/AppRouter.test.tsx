import React from 'react';
import renderer from 'react-test-renderer';
import ShallowRenderer from 'react-test-renderer/shallow';
import AppRouter, { AppRouterProps } from './AppRouter';
import Storage from '../StorageModel/Storage';
import { JustGraphitStore } from '../JustGraphitTypes';

const initialObject = { 'empty graph': [] } as unknown as JustGraphitStore;

const mockStorage: Partial<Storage> = {
  deleteObject: jest.fn(),
  getAllKeys: jest.fn(() => []),
  getInitialObject: () => initialObject,
  load: jest.fn(),
  save: jest.fn(),
  // persistentStore: mockLocalStorageWrapper,
  store: jest.fn() as unknown as JustGraphitStore,
  getAll: jest.fn(),
  getInitialKey: jest.fn(),
  updatePersistence: jest.fn(),
  isEmpty: jest.fn(),
};

jest.mock('../StorageModel/Storage', () => {
  return {
    Storage: jest.fn().mockImplementation(() => {
      return mockStorage;
    }),
  };
});

let appRouterProps: AppRouterProps;
beforeEach(() => {
  appRouterProps = {
    storage: mockStorage as unknown as Storage,
  };
});

test('deep snapshot', () => {
  const component = renderer.create(<AppRouter {...appRouterProps} />);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('shallow snapshot', () => {
  const renderer = ShallowRenderer.createRenderer();
  renderer.render(<AppRouter {...appRouterProps} />);
  const result = renderer.getRenderOutput();
  expect(result).toMatchSnapshot();
});
