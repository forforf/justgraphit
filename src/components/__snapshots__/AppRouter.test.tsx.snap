// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`deep snapshot 1`] = `
Array [
  .emotion-0 {
  background-color: #fff;
  color: rgba(0, 0, 0, 0.87);
  -webkit-transition: box-shadow 300ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: box-shadow 300ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  box-shadow: 0px 2px 4px -1px rgba(0,0,0,0.2),0px 4px 5px 0px rgba(0,0,0,0.14),0px 1px 10px 0px rgba(0,0,0,0.12);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  width: 100%;
  box-sizing: border-box;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  position: fixed;
  z-index: 1100;
  top: 0;
  left: auto;
  right: 0;
  background-color: #1976d2;
  color: #fff;
}

@media print {
  .emotion-0 {
    position: absolute;
  }
}

<header
    className="MuiPaper-root MuiPaper-elevation MuiPaper-elevation4 MuiAppBar-root MuiAppBar-colorPrimary MuiAppBar-positionFixed mui-fixed emotion-0"
    title="JustGraphIt"
  />,
  .emotion-0 {
  overflow: hidden;
  min-height: 48px;
  -webkit-overflow-scrolling: touch;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

@media (max-width:599.95px) {
  .emotion-0 .MuiTabs-scrollButtons {
    display: none;
  }
}

.emotion-1 {
  position: relative;
  display: inline-block;
  -webkit-flex: 1 1 auto;
  -ms-flex: 1 1 auto;
  flex: 1 1 auto;
  white-space: nowrap;
  overflow-x: hidden;
  width: 100%;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.emotion-3 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  position: relative;
  box-sizing: border-box;
  -webkit-tap-highlight-color: transparent;
  background-color: transparent;
  outline: 0;
  border: 0;
  margin: 0;
  border-radius: 0;
  padding: 0;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  vertical-align: middle;
  -moz-appearance: none;
  -webkit-appearance: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: inherit;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 500;
  font-size: 0.875rem;
  line-height: 1.25;
  letter-spacing: 0.02857em;
  text-transform: uppercase;
  max-width: 360px;
  min-width: 90px;
  position: relative;
  min-height: 48px;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  padding: 12px 16px;
  overflow: hidden;
  white-space: normal;
  text-align: center;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  color: rgba(0, 0, 0, 0.6);
}

.emotion-3::-moz-focus-inner {
  border-style: none;
}

.emotion-3.Mui-disabled {
  pointer-events: none;
  cursor: default;
}

@media print {
  .emotion-3 {
    -webkit-print-color-adjust: exact;
    color-adjust: exact;
  }
}

.emotion-3.Mui-selected {
  color: #1976d2;
}

.emotion-3.Mui-disabled {
  color: rgba(0, 0, 0, 0.38);
}

.emotion-4 {
  position: absolute;
  height: 2px;
  bottom: 0;
  width: 100%;
  -webkit-transition: all 300ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: all 300ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  background-color: #1976d2;
}

<div
    className="MuiTabs-root emotion-0"
  >
    <div
      className="MuiTabs-scroller MuiTabs-fixed emotion-1"
      onScroll={[Function]}
      style={
        Object {
          "marginBottom": -0,
          "overflow": "hidden",
        }
      }
    >
      <div
        aria-orientation={null}
        className="MuiTabs-flexContainer emotion-2"
        onKeyDown={[Function]}
        role="tablist"
      >
        <a
          aria-selected={true}
          className="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary Mui-selected emotion-3"
          href="#/"
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role="tab"
          tabIndex={0}
        >
          GraphIt
          <span
            className="MuiTabs-indicator emotion-4"
            style={Object {}}
          />
        </a>
        <a
          aria-selected={false}
          className="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary emotion-3"
          href="#/edit"
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role="tab"
          tabIndex={-1}
        >
          Editor
        </a>
        <a
          aria-selected={false}
          className="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary emotion-3"
          href="#/about"
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role="tab"
          tabIndex={-1}
        >
          About
        </a>
      </div>
    </div>
  </div>,
  .emotion-0 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  position: relative;
  min-width: 0;
  padding: 0;
  margin: 0;
  border: 0;
  vertical-align: top;
}

.emotion-1 {
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 400;
  font-size: 1rem;
  line-height: 1.4375em;
  letter-spacing: 0.00938em;
  color: rgba(0, 0, 0, 0.87);
  box-sizing: border-box;
  position: relative;
  cursor: text;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: relative;
  border-radius: 4px;
}

.emotion-1.Mui-disabled {
  color: rgba(0, 0, 0, 0.38);
  cursor: default;
}

.emotion-1:hover .MuiOutlinedInput-notchedOutline {
  border-color: rgba(0, 0, 0, 0.87);
}

@media (hover: none) {
  .emotion-1:hover .MuiOutlinedInput-notchedOutline {
    border-color: rgba(0, 0, 0, 0.23);
  }
}

.emotion-1.Mui-focused .MuiOutlinedInput-notchedOutline {
  border-color: #1976d2;
  border-width: 2px;
}

.emotion-1.Mui-error .MuiOutlinedInput-notchedOutline {
  border-color: #d32f2f;
}

.emotion-1.Mui-disabled .MuiOutlinedInput-notchedOutline {
  border-color: rgba(0, 0, 0, 0.26);
}

.emotion-2 {
  font: inherit;
  letter-spacing: inherit;
  color: currentColor;
  padding: 4px 0 5px;
  border: 0;
  box-sizing: content-box;
  background: none;
  height: 1.4375em;
  margin: 0;
  -webkit-tap-highlight-color: transparent;
  display: block;
  min-width: 0;
  width: 100%;
  -webkit-animation-name: mui-auto-fill-cancel;
  animation-name: mui-auto-fill-cancel;
  -webkit-animation-duration: 10ms;
  animation-duration: 10ms;
  padding: 16.5px 14px;
}

.emotion-2::-webkit-input-placeholder {
  color: currentColor;
  opacity: 0.42;
  -webkit-transition: opacity 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: opacity 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
}

.emotion-2::-moz-placeholder {
  color: currentColor;
  opacity: 0.42;
  -webkit-transition: opacity 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: opacity 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
}

.emotion-2:-ms-input-placeholder {
  color: currentColor;
  opacity: 0.42;
  -webkit-transition: opacity 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: opacity 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
}

.emotion-2::-ms-input-placeholder {
  color: currentColor;
  opacity: 0.42;
  -webkit-transition: opacity 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: opacity 200ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
}

.emotion-2:focus {
  outline: 0;
}

.emotion-2:invalid {
  box-shadow: none;
}

.emotion-2::-webkit-search-decoration {
  -webkit-appearance: none;
}

label[data-shrink=false]+.MuiInputBase-formControl .emotion-2::-webkit-input-placeholder {
  opacity: 0!important;
}

label[data-shrink=false]+.MuiInputBase-formControl .emotion-2::-moz-placeholder {
  opacity: 0!important;
}

label[data-shrink=false]+.MuiInputBase-formControl .emotion-2:-ms-input-placeholder {
  opacity: 0!important;
}

label[data-shrink=false]+.MuiInputBase-formControl .emotion-2::-ms-input-placeholder {
  opacity: 0!important;
}

label[data-shrink=false]+.MuiInputBase-formControl .emotion-2:focus::-webkit-input-placeholder {
  opacity: 0.42;
}

label[data-shrink=false]+.MuiInputBase-formControl .emotion-2:focus::-moz-placeholder {
  opacity: 0.42;
}

label[data-shrink=false]+.MuiInputBase-formControl .emotion-2:focus:-ms-input-placeholder {
  opacity: 0.42;
}

label[data-shrink=false]+.MuiInputBase-formControl .emotion-2:focus::-ms-input-placeholder {
  opacity: 0.42;
}

.emotion-2.Mui-disabled {
  opacity: 1;
  -webkit-text-fill-color: rgba(0, 0, 0, 0.38);
}

.emotion-2:-webkit-autofill {
  -webkit-animation-duration: 5000s;
  animation-duration: 5000s;
  -webkit-animation-name: mui-auto-fill;
  animation-name: mui-auto-fill;
}

.emotion-2:-webkit-autofill {
  border-radius: inherit;
}

.emotion-3 {
  text-align: left;
  position: absolute;
  bottom: 0;
  right: 0;
  top: -5px;
  left: 0;
  margin: 0;
  padding: 0 8px;
  pointer-events: none;
  border-radius: inherit;
  border-style: solid;
  border-width: 1px;
  overflow: hidden;
  min-width: 0%;
  border-color: rgba(0, 0, 0, 0.23);
}

.emotion-4 {
  float: unset;
  padding: 0;
  line-height: 11px;
  -webkit-transition: width 150ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
  transition: width 150ms cubic-bezier(0.0, 0, 0.2, 1) 0ms;
}

.emotion-10 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
  position: relative;
  box-sizing: border-box;
  -webkit-tap-highlight-color: transparent;
  background-color: transparent;
  outline: 0;
  border: 0;
  margin: 0;
  border-radius: 0;
  padding: 0;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  vertical-align: middle;
  -moz-appearance: none;
  -webkit-appearance: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: inherit;
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 500;
  font-size: 0.875rem;
  line-height: 1.75;
  letter-spacing: 0.02857em;
  text-transform: uppercase;
  min-width: 64px;
  padding: 6px 16px;
  border-radius: 4px;
  -webkit-transition: background-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,box-shadow 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,border-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  transition: background-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,box-shadow 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,border-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;
  color: #fff;
  background-color: #1976d2;
  box-shadow: 0px 3px 1px -2px rgba(0,0,0,0.2),0px 2px 2px 0px rgba(0,0,0,0.14),0px 1px 5px 0px rgba(0,0,0,0.12);
}

.emotion-10::-moz-focus-inner {
  border-style: none;
}

.emotion-10.Mui-disabled {
  pointer-events: none;
  cursor: default;
}

@media print {
  .emotion-10 {
    -webkit-print-color-adjust: exact;
    color-adjust: exact;
  }
}

.emotion-10:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
  background-color: #1565c0;
  box-shadow: 0px 2px 4px -1px rgba(0,0,0,0.2),0px 4px 5px 0px rgba(0,0,0,0.14),0px 1px 10px 0px rgba(0,0,0,0.12);
}

@media (hover: none) {
  .emotion-10:hover {
    background-color: #1976d2;
  }
}

.emotion-10:active {
  box-shadow: 0px 5px 5px -3px rgba(0,0,0,0.2),0px 8px 10px 1px rgba(0,0,0,0.14),0px 3px 14px 2px rgba(0,0,0,0.12);
}

.emotion-10.Mui-focusVisible {
  box-shadow: 0px 3px 5px -1px rgba(0,0,0,0.2),0px 6px 10px 0px rgba(0,0,0,0.14),0px 1px 18px 0px rgba(0,0,0,0.12);
}

.emotion-10.Mui-disabled {
  color: rgba(0, 0, 0, 0.26);
  box-shadow: none;
  background-color: rgba(0, 0, 0, 0.12);
}

<div
    className="Graphit"
  >
    <div
      className="wrapper"
    >
      <div
        className="row"
      >
        <div
          className="left col-2-3"
        >
          <div
            className="GraphInput"
          >
            <div
              className="GraphInput-name"
            >
              <h3>
                empty graph
              </h3>
            </div>
            <form>
              <div
                className="MuiFormControl-root MuiTextField-root emotion-0"
                onKeyPress={[Function]}
              >
                <div
                  className="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl emotion-1"
                  onClick={[Function]}
                >
                  <input
                    aria-invalid={false}
                    autoFocus={false}
                    className="MuiOutlinedInput-input MuiInputBase-input emotion-2"
                    disabled={false}
                    onAnimationStart={[Function]}
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    placeholder="Name your graph"
                    required={false}
                    type="text"
                    value="empty graph"
                  />
                  <fieldset
                    aria-hidden={true}
                    className="MuiOutlinedInput-notchedOutline emotion-3"
                  >
                    <legend
                      className="emotion-4"
                    >
                      <span
                        className="notranslate"
                        dangerouslySetInnerHTML={
                          Object {
                            "__html": "&#8203;",
                          }
                        }
                      />
                    </legend>
                  </fieldset>
                </div>
              </div>
              <br />
              <div
                className="MuiFormControl-root MuiTextField-root emotion-0"
                onKeyPress={[Function]}
              >
                <div
                  className="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl emotion-1"
                  onClick={[Function]}
                >
                  <input
                    aria-invalid={false}
                    autoFocus={false}
                    className="MuiOutlinedInput-input MuiInputBase-input emotion-2"
                    disabled={false}
                    onAnimationStart={[Function]}
                    onBlur={[Function]}
                    onChange={[Function]}
                    onFocus={[Function]}
                    placeholder="Enter number to graph"
                    required={false}
                    type="number"
                    value=""
                  />
                  <fieldset
                    aria-hidden={true}
                    className="MuiOutlinedInput-notchedOutline emotion-3"
                  >
                    <legend
                      className="emotion-4"
                    >
                      <span
                        className="notranslate"
                        dangerouslySetInnerHTML={
                          Object {
                            "__html": "&#8203;",
                          }
                        }
                      />
                    </legend>
                  </fieldset>
                </div>
              </div>
              <br />
              <button
                className="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  emotion-10"
                disabled={false}
                onBlur={[Function]}
                onClick={[Function]}
                onContextMenu={[Function]}
                onDragLeave={[Function]}
                onFocus={[Function]}
                onKeyDown={[Function]}
                onKeyPress={[Function]}
                onKeyUp={[Function]}
                onMouseDown={[Function]}
                onMouseLeave={[Function]}
                onMouseUp={[Function]}
                onTouchEnd={[Function]}
                onTouchMove={[Function]}
                onTouchStart={[Function]}
                tabIndex={0}
                type="button"
              >
                Save
              </button>
            </form>
          </div>
          <div
            className="GraphChart"
          >
            <div>
              This is an EmptyGraph
            </div>
          </div>
        </div>
        <div
          className="right col-1-3"
        >
          <div
            className="GraphList"
          >
            <h3>
              Graph List
            </h3>
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;

exports[`shallow snapshot 1`] = `
<HashRouter>
  <ForwardRef(AppBar)
    title="JustGraphIt"
  />
  <NavBar />
  <Routes>
    <Route
      element={
        <Graphit
          addNewNumber={[Function]}
          changeSelectedGraph={[Function]}
          deleteGraph={[Function]}
          graphNameList={Array []}
          graphObject={
            GraphObject {
              "data": Array [],
              "name": "empty graph",
            }
          }
        />
      }
      path="/"
    />
    <Route
      element={<About />}
      path="/about"
    />
    <Route
      element={
        <Editor
          graphObject={
            GraphObject {
              "data": Array [],
              "name": "empty graph",
            }
          }
          saveToStorage={[MockFunction]}
          setGraphObject={[Function]}
        />
      }
      path="/edit"
    />
  </Routes>
</HashRouter>
`;
